adapters/sspbc/sspbc.go

```diff
@@ -0,0 +1,407 @@
+	}
+
+	bannerTemplate, err := template.ParseFiles(pathToTemplate)
+	if err != nil {
+		return nil, err
+	}
+
+	bidder := &SspbcAdapter{
+		endpoint:       config.Endpoint,
+		version:        version,
+		bannerTemplate: bannerTemplate,
+	}
+
+	return bidder, nil
+}
+
+func (a *SspbcAdapter) MakeRequests(request *openrtb2.BidRequest, requestInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	var errors []error
+
+	formattedRequest, err := formatSsbcRequest(a, request)
+	if err != nil {
+		glog.Errorf("SSPBC: cannot prepare request")
+		errors = append(errors, err)
+		return nil, errors
+	}
+
+	requestJSON, err := json.Marshal(formattedRequest)
+	if err != nil {
+		glog.Errorf("SSPBC: cannot marshal request")
+		errors = append(errors, err)
+		return nil, errors
+	}
+
+	requestData := &adapters.RequestData{
+		Method: "POST",
+		Uri:    fmt.Sprintf("%s?bdver=%s&inver=0", a.endpoint, a.version),
+		Body:   requestJSON,
+	}
+
+	return []*adapters.RequestData{requestData}, nil
+}
+
+func (a *SspbcAdapter) MakeBids(internalRequest *openrtb2.BidRequest, externalRequest *adapters.RequestData, externalResponse *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	/*
+		  proxy responds with the following format
+			{
+			"cur": "PLN",
+			"id": "...",
+			"seatbid": [
+				{
+					"bid": [
+						{
+							"adm": "....",
+							"adomain": [
+								"sspbc-test"
+							],
+							"crid": "1234",
+							"ext": {
+								"adlabel": "Reklama",
+								"pubid": "431",
+								"siteid": "237503",
+								"slotid": "005",
+								"tagid": "slot"
+							},
+							"h": 250,
+						"id": "...",
+								"impid": "005",
+								"price": 95.95,
+							"w": 300
+						}
+					],
+					"seat": "sspbc-test"
+				}
+			],
+			"sn": "sspbc-test"
+			}
+
+		Note - we cannot read site SN, since response.sn is not defined in
+		openRTB2.BidResponse structure
+
+		For now we set SN as sspbc_go
+
+		Long term SN should be returned in bid.ext
+	*/
+
+	var errors []error
+
+	if externalResponse.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if externalResponse.StatusCode != http.StatusOK {
+		err := &errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d.", externalResponse.StatusCode),
+		}
+		return nil, []error{err}
+	}
+
+	var response openrtb2.BidResponse
+	if err := json.Unmarshal(externalResponse.Body, &response); err != nil {
+		return nil, []error{err}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(len(internalRequest.Imp))
+	bidResponse.Currency = response.Cur
+
+	for _, seatBid := range response.SeatBid {
+		for _, bid := range seatBid.Bid {
+			var BidType openrtb_ext.BidType
+			var BidId = bid.ImpID
+
+			/*
+			  here we should make a call to getBidType method, and based on detected type
+			  make call to createBannerAd, createVideoAd, createNativeAd methods
+
+			  for now we set it to "banner"
+			*/
+			BidType = openrtb_ext.BidTypeBanner
```

Comment:
We will accept a hardcoded banner value since the yaml file only defines a banner type. If you change that in the future, we'll ask for a proper implementation here. Ideally, the bid response should include the creative media type and we will soon add support for the new mtype field added in OpenRTB 2.6.

Reply:
Added basic detection of bid type (at the moment only for only banner bids). Bids that are not recognized will not be appended to response.